# 01. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing data sets

In [2]:
# Importing df_prods
path = r'C:\Users\veren\OneDrive\Desktop\Data Analytics\26-10-2021 Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\veren\\OneDrive\\Desktop\\Data Analytics\\26-10-2021 Instacart Basket Analysis'

In [4]:
df_prods = 

In [5]:
# Checking with df.columns-function
df_prods.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
# Checking with df.columns-function
df_ords.columns

Index(['Unnamed: 0', 'order_id', 'user_id', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [8]:
# Consistency check on df. prods with df.describe()
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [9]:
# Consistency check on df_ ords with df.describe
df_ords.describe() 

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [10]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [11]:
df_ords['order_id'].dtype

dtype('int64')

In [12]:
# I am assigning a string value to order_id again
df_ords['order_id']=df_ords['order_id'].astype('str')

In [13]:
# Checking
df_ords['order_id'].dtype

dtype('O')

In [14]:
# Checking datatype of user_id
df_ords['user_id'].dtype

dtype('int64')

In [15]:
# I am assigning a string value to order_id again
df_ords['user_id']=df_ords['user_id'].astype('str')

In [16]:
# Checking datatype of user_id again
df_ords['user_id'].dtype

dtype('O')

# 03. Checking for mixed-type data

In [17]:
# Create a dataframe
df_test=pd.DataFrame()

In [18]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [19]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [20]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [21]:
# Converting df_test into string values
df_test['mix'] = df_test['mix'].astype('str')

In [22]:
df_test['mix'].dtype

dtype('O')

# 04. Missing values

In [23]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [24]:
# Create a subset with the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [25]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [26]:
# Comparing the number of rows in your current dataframe with the number in the subset once the missing rows have been removed
df_prods.shape

(49693, 5)

In [27]:
# Creating the new dataframe df_prods_clean without null values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [28]:
# Running the shape function again to see if the number of rows has decreased
df_prods_clean.shape

(49677, 5)

Success!

In [29]:
# Looking for duplicates in df_prods_clean making a new subset df_dups
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [30]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [31]:
# Checking the rows in df_prods_clean before dropping the duplicates
df_prods_clean.shape

(49677, 5)

In [32]:
# Deleting duplicates and saving the df without duplicates in a new dataframe 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
# Checking the rows in df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

Success!

# 05. Task

In [34]:
# 2. Run the df.describe() function on your df_prods dataframe. 
#    Using your new knowledge about how to interpret the output of this function, 
#    share in a markdown cell whether anything about the data looks off or should be investigated further.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum value in the price column seems pretty high. Since Instacart is a "normal" supermarket and doesn't have luxury goods, the values of the price column should be further investigated.

In [36]:
#3. Check for mixed-type data in your df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Python didn't print anything, so there are no mixed-type data columns in df_ords.

In [37]:
# 5. Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The column days_since_prior_order has 206209 missing values. This probably is the case because some customers have only placed one order which means there is no entry in the days_since_prior_order column. So these missing values are actually unique and necessary data points.

In [38]:
# 6. Address the missing values using an appropriate method
# Method: Assigning a new dataframe with a new column that acts as a flag
df_ords_clean = df_ords

In [39]:
df_ords_clean['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [40]:
df_ords_clean.head()

Unnamed: 0 order_id user_id  order_number  orders_day_of_week  \
0           0  2539329       1             1                   2   
1           1  2398795       1             2                   3   
2           2   473747       1             3                   3   
3           3  2254736       1             4                   4   
4           4   431534       1             5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

Explaining the method: It was not possible to delete the missing values because they provide important information i.e. that it is the customer's first order. For the same reason, it is not possible to impute the missing data. Therefore, we need to create a new column with variables that acts as a flag based on the missing value. A good name for this new column is first_order and indicate whether it is the customers first order or not (True or False). This column then explains the missing values in the days_since_prior_order column.

In [41]:
# 7. Run a check for duplicate values in your df_ords data
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [42]:
df_ords_dups.head()

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

There are no duplicate values in df_ords_clean

In [43]:
# 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder 
#    and give them appropriate, succinct names.
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [44]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))